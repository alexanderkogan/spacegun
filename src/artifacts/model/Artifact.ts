export interface Artifact {
    name: string
    data: object
}
