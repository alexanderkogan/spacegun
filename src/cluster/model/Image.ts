export interface Image {
    url: string
    name: string
    tag: string
}
