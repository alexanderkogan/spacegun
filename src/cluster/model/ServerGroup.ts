export interface ServerGroup {
    cluster: string
    namespace?: string
}
