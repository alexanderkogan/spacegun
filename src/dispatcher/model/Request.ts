import { RequestInput } from "./RequestInput"

export interface Request<Input, Output> {
    module: string
    procedure: string
    input?: (input: Input | undefined) => RequestInput
    mapper?: (input: RequestInput) => Input | undefined
    output?: (output: Output | undefined) => object
}
