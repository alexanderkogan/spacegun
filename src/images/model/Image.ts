import { Tag } from './Tag'

export interface Image {
    url: string
    name: string
    tag: Tag
}
