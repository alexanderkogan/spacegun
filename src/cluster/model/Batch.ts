import { Image } from './Image'

export interface Batch {
    name: string
    image?: Image
}
