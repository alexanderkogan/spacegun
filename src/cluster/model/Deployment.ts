import { Image } from "@/cluster/model/Image"

export interface Deployment {
    name: string
    image?: Image
}
