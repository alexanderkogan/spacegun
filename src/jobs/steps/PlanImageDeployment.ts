import { IO } from "../../IO"

import { call } from "../../dispatcher"

import * as imageModule from "../../images/ImageModule"

import { DeploymentPlan } from "../model/DeploymentPlan"
import { Deployment } from "../../cluster/model/Deployment"
import { ServerGroup } from "../../cluster/model/ServerGroup"
import { Filter, matchesDeployment, matchesServerGroup } from "../model/Filter"
import { JobPlan } from "../model/JobPlan";

export type FetchedDeployment = {}

export class PlanImageDeployment {

    public constructor(
        readonly name: string,
        readonly tag: string | undefined,
        readonly semanticTagExtractor: RegExp | string | undefined,
        readonly filter?: Partial<Filter>,
        readonly io: IO = new IO()
    ) { }

    public async plan(group: ServerGroup, name: string, targetDeployments: Deployment[]): Promise<JobPlan> {
        if (!matchesServerGroup(this.filter, group)) {
            return { name, deployments: [] }
        }

        const deployments: DeploymentPlan[] = []
        for (const targetDeployment of targetDeployments) {
            if (!matchesDeployment(this.filter, targetDeployment)) {
                continue
            }
            this.io.out(`planning image deployment ${targetDeployment.name} in ${this.name}`)
            if (targetDeployment.image === undefined) {
                this.io.error(`${targetDeployment.name} in cluster ${group.cluster} has no image, so spacegun cannot determine the right image source`)
                continue
            }

            let tag = this.tag
            if (tag === undefined) {
                const tags = await call(imageModule.tags)(targetDeployment.image)
                tag = this.getNewestTag(tags)
            }

            if (tag !== undefined) {
                const image = await call(imageModule.image)({
                    tag, name: targetDeployment.image.name
                })

                if (targetDeployment.image.url !== image.url) {
                    deployments.push({
                        group,
                        image,
                        deployment: targetDeployment,
                    })
                }
            } else {
                throw new Error(`Could not find a tag for deployment ${targetDeployment.name} in cluster ${group.cluster}, namespace ${group.namespace}`)
            }
        }
        return { name, deployments }
    }

    private getNewestTag(tags: string[]): string | undefined {
        let sortableTags: string[]
        if (this.semanticTagExtractor !== undefined) {
            sortableTags = tags.map(tag => tag.match(this.semanticTagExtractor!))
                .filter(match => match !== null && match.length > 0)
                .map(match => match![0])
        } else {
            sortableTags = tags
        }

        sortableTags.sort((a, b) => b.localeCompare(a))
        if (sortableTags.length === 0) {
            return undefined
        }
        if (sortableTags.length > 1 && sortableTags[0] === sortableTags[1]) {
            return undefined
        }
        return sortableTags.length > 0 ? sortableTags[0] : undefined
    }
}
