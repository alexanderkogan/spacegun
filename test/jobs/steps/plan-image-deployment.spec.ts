import { Request } from "../../../src/dispatcher/model/Request"

import { PlanImageDeployment } from "../../../src/jobs/steps/PlanImageDeployment"
import { ServerGroup } from "../../../src/cluster/model/ServerGroup"
import { Deployment } from "../../../src/cluster/model/Deployment"
import { Image } from "../../../src/images/model/Image"

let mockImages: { [name: string]: Image } = {}
let mockTags: { [name: string]: string[] } = {}

const mockImageRequest = jest.fn()

jest.mock("../../../src/dispatcher/index", () => ({
    get: jest.fn(),
    call: (request: Request<any, any>) => {
        switch (request.module) {
            case "images": {
                switch (request.procedure) {
                    case "image": {
                        return (input: { name: string, tag: string }) => {
                            mockImageRequest(input)
                            return Promise.resolve(mockImages[input.name])
                        }
                    }
                    case "tags": {
                        return (input: { name: string }) => Promise.resolve(mockTags[input.name])
                    }
                }
                break
            }
        }
        return undefined
    },
    add: () => { },
    path: () => ""
}))

describe(PlanImageDeployment.name, () => {

    beforeEach(() => {
        mockImageRequest.mockReset()
    })

    const group: ServerGroup = { cluster: "targetCluster", namespace: "namespace" }
    const step = new PlanImageDeployment("name", "latest", undefined, {
        namespaces: ["namespace"],
        deployments: ["deployment1"]
    })
    const targetDeployments: Deployment[] = [{
        name: "deployment1", image: { name: "image1", url: "url2" }
    }]

    it("plans deployments", async () => {
        // given
        mockImages = { image1: { name: "image1", tag: "tag", url: "url1" } }

        // when
        const plan = await step.plan(group, "pipeline1", targetDeployments)

        // then
        expect(plan.deployments).toEqual([
            {
                deployment: { name: "deployment1", image: { name: "image1", url: "url2" } },
                group: { cluster: "targetCluster", namespace: "namespace" },
                image: { name: "image1", tag: "tag", url: "url1" }
            }
        ])
    })

    it("ignores deployments that have same source image", async () => {
        // given
        mockImages = { image1: { name: "image1", tag: "tag", url: "url2" } }

        // when
        const plan = await step.plan(group, "pipeline1", targetDeployments)

        // then
        expect(plan.deployments).toEqual([])
    })

    it("ignores deployments that do not match", async () => {
        // given
        mockImages = { image1: { name: "image1", tag: "tag", url: "url1" } }

        // when
        const plan = await step.plan(group, "pipeline1", [{ name: "deployment2", image: { name: "image1", url: "url2" } }])

        // then
        expect(plan.deployments).toEqual([])
    })

    it("ignores deployments that have not image", async () => {
        // given
        step.io.error = jest.fn()
        mockImages = { image1: { name: "image1", tag: "tag", url: "url1" } }

        // when
        const plan = await step.plan(group, "pipeline1", [{ name: "deployment1" }])

        // then
        expect(plan.deployments).toEqual([])
        expect(step.io.error).toHaveBeenCalledWith(
            "deployment1 in cluster targetCluster has no image, so spacegun cannot determine the right image source"
        )
    })

    it("does not plan if server group does not match", async () => {
        // given
        mockImages = { image1: { name: "image1", tag: "tag", url: "url1" } }

        // when
        const plan = await step.plan({ cluster: "cluster", namespace: "other" }, "pipeline1", targetDeployments)

        // then
        expect(plan.deployments).toEqual([])
    })

    describe("tag resolution", () => {

        const latestImage = { name: "image1", tag: "latest", url: "url1" }

        it("uses the defined tag if present", async () => {
            // given
            mockImages = { image1: latestImage }
            const step = new PlanImageDeployment("name", "latest", undefined, undefined)

            // when
            const plan = await step.plan(group, "pipeline1", targetDeployments)

            // then
            expect(plan.deployments[0].image).toEqual(latestImage)
        })

        it("uses the lexicographically largest tag", async () => {
            // given
            mockTags = { image1: ["a", "b", "c"] }
            mockImages = { image1: latestImage }
            const step = new PlanImageDeployment("name", undefined, undefined, undefined)

            // when
            const plan = await step.plan(group, "pipeline1", targetDeployments)

            // then
            expect(mockImageRequest).toHaveBeenCalledWith({ name: "image1", tag: "c" })
            expect(plan.deployments[0].image).toEqual(latestImage)
        })

        it("uses the matching tag", async () => {
            // given
            mockTags = { image1: ["someTag", "latest"] }
            mockImages = { image1: latestImage }
            const step = new PlanImageDeployment("name", undefined, "latest", undefined)

            // when
            const plan = await step.plan(group, "pipeline1", targetDeployments)

            // then
            expect(mockImageRequest).toHaveBeenCalledWith({ name: "image1", tag: "latest" })
            expect(plan.deployments[0].image).toEqual(latestImage)
        })

        it("extracts the matching part and uses the lexicographically largest one", async () => {
            // given
            mockTags = { image1: ["latest1", "latest2", "latest3", "latest4"] }
            mockImages = { image1: latestImage }
            const step = new PlanImageDeployment("name", undefined, "latest.", undefined)

            // when
            const plan = await step.plan(group, "pipeline1", targetDeployments)

            // then
            expect(mockImageRequest).toHaveBeenCalledTimes(1)
            expect(mockImageRequest).toHaveBeenCalledWith({ name: "image1", tag: "latest4" })
            expect(plan.deployments[0].image).toEqual(latestImage)
        })


        it("works with more complex regex", async () => {
            mockTags = { image1: ["ztag-2003-1-92", "aaatag-2020-10-01", "aaatag-2018-99-99"] }
            mockImages = { image1: latestImage }
            const step = new PlanImageDeployment("name", undefined, "\\d{4}\-\\d{1,2}\-\\d{1,2}$", undefined)

            // when
            await step.plan(group, "pipeline1", targetDeployments)

            // then
            expect(mockImageRequest).toHaveBeenCalledTimes(1)
            expect(mockImageRequest).toHaveBeenCalledWith({ name: "image1", tag: "aaatag-2020-10-01" })
        })

        it("throws an error if it cannot match a single tag", async () => {
            // given
            mockTags = { image1: ["notLatest", "otherTag", "coolTag"] }
            const step = new PlanImageDeployment("name", undefined, "latest.", undefined)

            // when /
            expect(step.plan(group, "pipeline1", targetDeployments)).rejects.toMatchSnapshot()
        })

        it("throws an error if it cannot match a unique tag", async () => {
            // given
            mockTags = { image1: ["latest1", "latest2"] }
            const step = new PlanImageDeployment("name", undefined, "latest", undefined)

            // when / then
            expect(step.plan(group, "pipeline1", targetDeployments)).rejects.toMatchSnapshot()
        })
    })
})
