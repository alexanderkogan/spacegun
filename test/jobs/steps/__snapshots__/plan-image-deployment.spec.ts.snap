// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlanImageDeployment tag resolution throws an error if it cannot match a single tag 1`] = `[Error: Could not find a tag for deployment deployment1 in cluster targetCluster, namespace namespace]`;

exports[`PlanImageDeployment tag resolution throws an error if it cannot match a unique tag 1`] = `[Error: Could not find a tag for deployment deployment1 in cluster targetCluster, namespace namespace]`;
